File: tests/integration/i_simpleapi.c

#include "../../src/http.h"
#include "../../src/server.h"
#include <string.h>

void simple_callback(Response* res, Request* req)
{
    char json[] = "{\"test\": \"callback\"}";
    JSONString* jstring = copy_chars(json, strlen(json));
    JSONObject* obj = parse_json(jstring, NULL);
    send_json(res, obj);
    free_json(obj);
    STRINGP_FREE(jstring);
}

#include <stdio.h>
void data_callback(Response* res, Request* req)
{
    bool success;
    JSONObject* json_obj = parse_json(&req->content, &success);
    printf("JSON content: %s\n", req->content.chars);
    char* json = NULL;
    if (success == false) {
        json = "{\"error\": \"Parse failed!\"}";
    } else {
        String* tmp = json_get_string_c(json_obj, "tdata");
        if (strcmp(tmp->chars, "test1") == 0) {
            json = "{\"result\": \"test1\"}";
        } else if (strcmp(tmp->chars, "test2") == 0) {
            json = "{\"result\": \"test2\"}";
        } else {
            json = "{\"result\": \"not found\"}";
        }
        STRINGP_FREE(tmp);
    }
    printf("JSON: %s\n", json);
    JSONString* jstring = copy_chars(json, strlen(json));
    JSONObject* obj = parse_json(jstring, NULL);
    send_json(res, obj);
    free_json(obj);
    STRINGP_FREE(jstring);
}

void parameter_callback(Response* res, Request* req)
{
    JSONString* json_str = json_get_string_c(req->params, "param");
    char* json = NULL;
    if (json_str == NULL) {
        json = "{\"error\": \"No 'param' parameter found!\"}";
    } else {
        if (strcmp(json_str->chars, "1") == 0) {
            json = "{\"result\": \"1\"}";
        } else if (strcmp(json_str->chars, "2") == 0) {
            json = "{\"result\": \"2\"}";
        } else {
            json = "{\"result\": \"not found\"}";
        }
    }
    JSONString* jstring = copy_chars(json, strlen(json));
    JSONObject* obj = parse_json(jstring, NULL);
    send_json(res, obj);
    free_json(obj);
    STRINGP_FREE(jstring);
}

int main(int argc, char const* argv[])
{

    RestServer rs;
    init_server(&rs);
    add_url(&rs, "/", simple_callback);
    add_url(&rs, "/api", data_callback);
    add_url(&rs, "/parameter/:param", parameter_callback);
    return run_server(&rs);
}

File: tests/integration/i_simpleapi.py
import json
import unittest
import urllib.request as re

server = "http://localhost:8888"

class TestSimpleCallback(unittest.TestCase):

    def test_simple(self):
        req = re.Request(url=f"{server}/", method='GET')
        with re.urlopen(req) as f:
            data = f.read()
            self.assertNotEqual(data, None)
            j = json.loads(data)
            self.assertEqual(j['test'], 'callback')

    def test_parameter(self):
        req = re.Request(url=f"{server}/parameter/1")
        with re.urlopen(req) as f:
            data = f.read()
            self.assertNotEqual(data, None)
            self.assertNotIn('error', str(data))
            j = json.loads(data)
            self.assertEqual(j['result'], '1')
        req = re.Request(url=f"{server}/parameter/2")
        with re.urlopen(req) as f:
            data = f.read()
            self.assertNotEqual(data, None)
            self.assertNotIn('error', str(data))
            j = json.loads(data)
            self.assertEqual(j['result'], '2')
        req = re.Request(url=f"{server}/parameter/3")
        with re.urlopen(req) as f:
            data = f.read()
            self.assertNotEqual(data, None)
            self.assertNotIn('error', str(data))
            j = json.loads(data)
            self.assertEqual(j['result'], 'not found')

    def test_data(self):
        req = re.Request(url=f"{server}/api", method='POST', data=b'{"tdata": "test1"}')
        with re.urlopen(req) as f:
            data = f.read()
            self.assertNotEqual(data, None)
            self.assertNotIn('error', str(data))
            j = json.loads(data)
            self.assertEqual(j['result'], 'test1')
        req = re.Request(url=f"{server}/api", method='POST', data=b'{"tdata": "test2"}')
        with re.urlopen(req) as f:
            data = f.read()
            self.assertNotEqual(data, None)
            self.assertNotIn('error', str(data))
            j = json.loads(data)
            self.assertEqual(j['result'], 'test2')
        req = re.Request(url=f"{server}/api", method='POST', data=b'{"tdata": "test3"}')
        with re.urlopen(req) as f:
            data = f.read()
            self.assertNotEqual(data, None)
            self.assertNotIn('error', str(data))
            j = json.loads(data)
            self.assertEqual(j['result'], 'not found')



if __name__ == '__main__':
        unittest.main()
